flexget/components/hook/hook.py

```diff
@@ -0,0 +1,336 @@
+            config['stages'] = stages
+
+            tasks = config.get('tasks', [])
+            tasks = tasks if isinstance(tasks, list) else [tasks]
+            config['tasks'] = tasks
+
+            plugins = config.get('plugins', [])
+            plugins = plugins if isinstance(plugins, list) else [plugins]
+            config['plugins'] = plugins
+
+            when = config.get('when', [])
+            when = when if isinstance(when, list) else [when]
+            config['when'] = when
+
+            phases = config.get('phases', [])
+            phases = phases if isinstance(phases, list) else [phases]
+            config['phases'] = phases
+
+            via = config.get('via', [])
+            via = via if isinstance(via, list) else [via]
+            config['via'] = via
+
+        Hook.config[task.id] = configs
+
+    @staticmethod
+    def get_config(task: Task):
+        if Hook.config is None or not task.id in Hook.config:
+            return None
+
+        return Hook.config[task.id]
+
+    @staticmethod
+    def delete_config(task: Task):
+        if Hook.config is None or not task.id in Hook.config:
+            return
+
+        del Hook.config[task.id]
+
+    @staticmethod
+    def _is_builtins(plugin_check):
+        return plugin.plugins[plugin_check].builtin
+
+    # First Tasks
+    def on_task_prepare(self, task, config):
+        Hook.set_config(task, config)
+        Hook.on_phase_ended(task, task.current_phase)
+
+    def on_task_start(self, task, config):
+        Hook.set_config(task, config)
+        Hook.on_task_started(task)
+
+    @event('task.execute.started')
+    def on_task_started(self):
+        config = Hook.get_config(self)
+        if not config:
+            return
+
+        Hook.send_hook(
+            None,
+            task_to_dict(self),
+            config,
+            event=ENDPOINT['EVENT_TASK'],
+            stage=ENDPOINT['STAGE_START'],
+            name=getattr(self, 'name'),
+            task=self,
+        )
+
+    @event('task.execute.completed')
+    def on_task_end(self):
+        config = Hook.get_config(self)
+        Hook.delete_config(self)
+        if not config:
+            return
+
+        Hook.send_hook(
+            None,
+            task_to_dict(self),
+            config,
+            event=ENDPOINT['EVENT_TASK'],
+            stage=ENDPOINT['STAGE_END'],
+            name=getattr(self, 'name'),
+            task=self,
+        )
+
+    @event('task.execute.before_plugin')
+    def on_plugin_started(self, plugin_e):
+        config = Hook.get_config(self)
+        if not config:
+            return
+
+        if Hook._is_builtins(plugin_e):
+            return
+
+        Hook.send_hook(
+            None,
+            task_to_dict(self),
+            config,
+            evet=ENDPOINT['EVENT_PLUGIN'],
+            stage=ENDPOINT['STAGE_START'],
+            name=plugin_e,
+            task=self,
+        )
+
+    @event('task.execute.after_plugin')
```

Comment:
You're repeating this same code block for multiple times. You could have done a factory function and avoid repeating. But even better would be to extend our event dispatcher to have a wildcard hook and hook that.

Someting like

```
@event(event.ALL)
def on_event(self, event_name, ...):
   # generic logic here
```

Reply:
Would this hook all the events? For example the events of the plugin register? Would we intended to do that? Task envets, plugin event's, and phases event's are useful! But what other events would this trigger?